src/gsSolver/gsParameterDomainPreconditioners.hpp

```diff
@@ -0,0 +1,438 @@
+    // We do not provide:
+    //   dirichlet::penalize
+    //   dirichlet::nitsche
+    //   dirichlet::eliminatNormal
+
+}
+
+namespace {
+
+template <typename T>
+int estimateNonzerosPerRow(const gsBasis<T>& basis)
+{
+    int nnz = 1;
+    for (int i = 0; i < basis.dim(); ++i)
+        nnz *= 2 * basis.degree(i) + 1;
+    return nnz;
+}
+
+template <typename T>
+void localToGlobal(const gsMatrix<T>& localMat, const gsMatrix<unsigned>& localDofs, gsSparseMatrix<T>& globalMat)
+{
+    const int numActive = localDofs.rows();
+
+    for (index_t i = 0; i < numActive; ++i)
+    {
+        const int ii = localDofs(i,0);
+        for (index_t j = 0; j < numActive; ++j)
+        {
+            const int jj = localDofs(j,0);
+            globalMat.coeffRef(ii, jj) += localMat(i, j);
+        }
+    }
+}
+
+} // anonymous namespace
+
+template<typename T>
+gsSparseMatrix<T> gsParameterDomainPreconditioners<T>::assembleMass(const gsBasis<T>& basis)
+{
+    const int n = basis.size();
+
+    gsSparseMatrix<T> M(n, n);
+    M.reserve( gsVector<int>::Constant(n, estimateNonzerosPerRow(basis)) );
+
+    gsMatrix<T> localMass;
+
+    // Make domain element iterator
+    typename gsBasis<T>::domainIter domIt = basis.makeDomainIterator();
+
+    // Set the number of integration points for each element
+    gsVector<index_t> numQuadNodes( basis.dim() );
+    for (int i = 0; i < basis.dim(); ++i)
+        numQuadNodes[i] = basis.degree(i) + 1;
+    domIt->computeQuadratureRule( numQuadNodes );
+
+    // Start iteration over elements
+    for (; domIt->good(); domIt->next())
+    {
+        const index_t numActive = domIt->computeActiveFunctions().rows();
+        domIt->evaluateBasis( 0 );
+
+        localMass.setZero(numActive, numActive);
+
+        for (index_t k = 0; k < domIt->numQuNodes(); ++k)      // loop over quadrature nodes
+        {
+            const T weight = domIt->quWeights[k];
+            localMass.noalias() += weight * (domIt->basisValues().col(k) * domIt->basisValues().col(k).transpose());
+        }
+
+        localToGlobal(localMass, domIt->activeFuncs, M);
+    }
+
+    M.makeCompressed();
+
+    return M;
+}
+
+template<typename T>
+gsSparseMatrix<T> gsParameterDomainPreconditioners<T>::assembleStiffness(const gsBasis<T>& basis)
+{
+    const int n = basis.size();
+    const int dim = basis.dim();
+
+    gsSparseMatrix<T> K(n, n);
+    K.reserve( gsVector<int>::Constant(n, estimateNonzerosPerRow(basis)) );
+
+    gsMatrix<T> localStiffness; // (dense) stiffness matrix within one grid cell
+
+    // Make domain element iterator
+    typename gsBasis<T>::domainIter domIt = basis.makeDomainIterator();
+
+    // Set the number of integration points for each element
+    gsVector<index_t> numQuadNodes( dim );
+    for (int i = 0; i < basis.dim(); ++i)
+        numQuadNodes[i] = basis.degree(i) + 1;
+    domIt->computeQuadratureRule( numQuadNodes );
```

Comment:
domIt->computeQuadratureRule( numQuadNodes ); 
is deprecated

Reply:
i cannot find related information.